components/object_store/src/mem_cache.rs

```diff
@@ -0,0 +1,322 @@
+// Copyright 2022 CeresDB Project Authors. Licensed under Apache-2.0.
+
+//! An implementation of ObjectStore, which support
+//! 1. Cache based on memory, and support evict based on memory usage
+//! 2. Builtin Partition to reduce lock contention
+
+use std::{
+    collections::hash_map::DefaultHasher,
+    fmt::Display,
+    hash::{Hash, Hasher},
+    ops::Range,
+    sync::Arc,
+};
+
+use async_trait::async_trait;
+use bytes::Bytes;
+use futures::stream::BoxStream;
+use lru_weighted_cache::{LruWeightedCache, Weighted};
+use tokio::{io::AsyncWrite, sync::Mutex};
+use upstream::{path::Path, GetResult, ListResult, MultipartId, ObjectMeta, ObjectStore, Result};
+
+struct CachedBytes(Bytes);
+
+impl Weighted for CachedBytes {
+    fn weight(&self) -> usize {
+        self.0.len()
+    }
+}
+
+#[derive(Debug)]
+struct Partition {
+    inner: Mutex<LruWeightedCache<String, CachedBytes>>,
+}
+
+impl Partition {
+    fn new(mem_cap: usize) -> Self {
+        Self {
+            inner: Mutex::new(LruWeightedCache::new(1, mem_cap).expect("invalid params")),
+        }
+    }
+}
+impl Partition {
+    async fn get(&self, key: &String) -> Option<Bytes> {
+        let mut guard = self.inner.lock().await;
+        guard.get(key).map(|v| v.0.clone())
+    }
+
+    async fn insert(&self, key: String, value: Bytes) {
+        let mut guard = self.inner.lock().await;
+        // don't care error now.
+        _ = guard.insert(key, CachedBytes(value));
+    }
+}
+
+#[derive(Debug)]
+struct MemCache {
+    /// Max memory this store can use
+    mem_cap: usize,
+    partitions: Vec<Arc<Partition>>,
+    partition_mask: usize,
+}
+
+impl MemCache {
+    fn new(partition_bits: usize, mem_cap: usize) -> Self {
+        let partition_num = 1 << partition_bits;
+        let cap_per_part = mem_cap / partition_num;
+        let partitions = (0..partition_num)
+            .map(|_| Arc::new(Partition::new(cap_per_part)))
+            .collect::<Vec<_>>();
+
+        Self {
+            mem_cap,
+            partitions,
+            partition_mask: partition_num - 1,
+        }
+    }
+
+    fn locate_partition(&self, key: &String) -> Arc<Partition> {
```

Comment:
Why not make the `hasher` as a trait object and put it in `MemCache`. 
If so, it may be easy to do this  when we want a new hasher.

Reply:
&String to &str

Reply:
> it may be easy to do this when we want a new hasher.

I couldn't think of a situation to switch hasher, if we want, we can just change this here.

I don't want to expose a generic type here.

Reply:
> &String to &str

`&String` required by [get](https://github.com/jiacai2050/lru-weighted-cache/blob/master/src/lib.rs#L210)

Reply:
Maybe we can leave a comment to explain this.